src/user-status/userStatusReducer.js

```diff
@@ -21,35 +54,69 @@ export default (
     case ACCOUNT_SWITCH:
       return initialState;
 
-    case REGISTER_COMPLETE:
-      return action.data.user_status || initialState;
+    case REGISTER_COMPLETE: {
+      const { user_status } = action.data;
+      if (!user_status) {
+        // TODO(server-1.9.1): Drop this. We'll end up using
+        //   kBoringUserStatus to fill in missing data. That'll probably end
+        //   up OK, but it violates kBoringUserStatus's interface; see
+        //   jsdoc.
+        return initialState;
+      }
+
+      return action.data.user_status
+        ? Immutable.Map(
+            // N.B.: `user_status[user_id]` may be missing if the server thinks
+            //   its value would be too boring to send. We'll fill it in by
+            //   making kBoringUserStatus a fallback wherever we read the state.
+            //
+            //   …We could fill it in here, using the users in `action.data`.
+            //   But in large orgs like CZO, we'd risk lagging as we do that for
+            //   tens of thousands of users.
+            objectEntries(action.data.user_status).map(([id, status]) => [
+              makeUserId(Number.parseInt(id, 10)),
+              {
+                ...kBoringUserStatus,
+                ...status,
+              },
+            ]),
+          )
+        : initialState;
+    }
 
     case EVENT_USER_STATUS_UPDATE: {
-      const newUserStatus = { ...state[action.user_id] };
-      if (action.away !== undefined) {
-        if (action.away === true) {
-          newUserStatus.away = action.away;
-        } else {
-          delete newUserStatus.away;
-        }
+      const oldUserStatus =
+        state.get(action.user_id)
+        // The user had a boring status at /register time.
+        ?? kBoringUserStatus;
+
+      const newUserStatus = { ...oldUserStatus };
+      if (typeof action.away === 'boolean') {
+        newUserStatus.away = action.away;
+      }
+      if (typeof action.status_text === 'string') {
+        newUserStatus.status_text = action.status_text.length > 0 ? action.status_text : null;
+      }
```

Comment:
Oh! Somehow I missed this comment while working on the current revision. What do you think of this reasoning, quoting from my other comment https://github.com/zulip/zulip-mobile/pull/5224#discussion_r803162255:

> Incidentally, although either could work, I prefer `null` to `''`, mainly because Flow can't track separate types for empty and non-empty strings. Callers like the one in `UserStatusScreen` that want to use an empty string for their own UI choices, like clearing out an `<Input />`, can do so. But the model gives a nice and explicit `string` | `null`.

Reply:
I've just pushed the simplification to use `action.away ?? oldUserStatus.away` but haven't yet touched `status_text`; let me know what you think. 🙂

Reply:
Yeah, I like that reasoning for `null`: that it lets Flow track the difference between that and a value that's set. Continued below: https://github.com/zulip/zulip-mobile/pull/5224#discussion_r813550160